packages/grafana-data/src/utils/url.ts

```diff
@@ -102,31 +102,32 @@ function appendQueryToUrl(url: string, stringToAppend: string) {
   return url;
 }
 
+let prevSearch: string;
+let prevParams: UrlQueryMap;
+
 /**
  * Return search part (as object) of current url
  */
 function getUrlSearchParams(): UrlQueryMap {
-  const search = window.location.search.substring(1);
-  const searchParamsSegments = search.split('&');
-  const params: UrlQueryMap = {};
-  for (const p of searchParamsSegments) {
-    const keyValuePair = p.split('=');
-    if (keyValuePair.length > 1) {
-      // key-value param
-      const key = decodeURIComponent(keyValuePair[0]);
-      const value = decodeURIComponent(keyValuePair[1]);
-      if (key in params) {
-        params[key] = [...(params[key] as any[]), value];
-      } else {
-        params[key] = [value];
-      }
-    } else if (keyValuePair.length === 1) {
-      // boolean param
-      const key = decodeURIComponent(keyValuePair[0]);
-      params[key] = true;
+  const curSearch = window.location.search.substring(1);
+
+  if (curSearch !== prevSearch) {
+    prevParams = {};
+
+    const curParsed = new URLSearchParams(curSearch);
+
+    // @ts-ignore
+    for (const key of new Set(curParsed.keys())) {
+      // @ts-ignore
+      let all = curParsed.getAll(key);
+      // @ts-ignore
+      prevParams[key] = all.length === 1 && all[0] === '' ? true : all;
     }
```

Comment:
```suggestion
     // @ts-ignore
    for (const [key, value] of curParsed.entries()) {
      prevParams[key] = [...(prevParams[key] || []), value || true]
    }
```
Wouldn't it be faster to get all the entries so then we can skip the `getAll` step? I guess array spread could be quite expensive though, but maybe using smth like `concat` would help 🤔

Reply:
`curParsed.entries()` won't give you unique keys. so you'll get back things like `["a","1"],["a","2"]` for `"a=1&a=2`:

![image](https://user-images.githubusercontent.com/43234/128334096-51f53588-b97d-4834-bc0f-fcf0dcb7ea88.png)

Reply:
eh, nvm. sorry 5am here. i'm not sure which version is faster but with a cache it likely makes no practical difference. concat is probably faster than spread, but i dunno vs `Set()` + `getAll()`.

Reply:
at least `entries` is less lines then `new Set` + `getAll()` 😂

Reply:
touché